### ISSUE TITLE
Enable bake mode to build images

### ISSUE THREAD

**Issue by YevheniiSemendiak** [neuro-inc/neuro-flow#311]
Actually, this FR contains several requests:
1. Bake description file should be self-contained. We should be able to build images, which are defined in its description. (currently, we could build only images listed in the `live.yaml` file). However, the behavior should echo with a bake behavior - it should build all the images. The command name might be smth like `neuro-flow prepare <batch>`.

2. Add alias command, which will call `neuro-flow prepare <batch>` and `neuro-flow bake <batch>` in a row. For instance, `neuro-flow cook <batch>`.

**Comment by romasku** [comment]
I see the problem, but see some issues with the suggested approach:
- "Batch" mode idea is that you fire command and free to power-off your PC -- bake will continue (except for local actions). So that `cook` command will have to start some "meta-executor" job, that will first build images and the will run normal executor. And if the building of images fails, it will be unclear how to show an error.
- It's not convenient (as for me) to manually select a moment to "build the images". Users can forget to run, or run it more often than required. And then we will end up wanting to add caching here...

I think it's better to solve this using actions:
```
kind: batch
tasks:
  - build_image_1:
    action: gh:neuro-actions/build-image
    args:
      context: storage:path/to/context
      dockerfile: storage:path/to/Dockerfile
      build_args: build_args_here
      env: ${{ to_json({"ENV_1": "foo", "ENV_2": "bar"}) }}
      volumes: ${{ to_json(["storage:volume1", "storage:volume2") }} 
  - real_task:
    image: ${{ needs.build_image_1.image.ref }}
```
Action as above is already possible but has the limitation that context and Dockerfile should be uploaded manually (maybe by some auxiliary local action).

If we add support of local actions inside batch actions (only in case it does not depend directly/indirectly on any remote task), it will be possible to implement the following:
```
kind: batch
tasks:
  - build_image_1:
    action: gh:neuro-actions/build-image
    args:
      context: ${{ flow.workspace / 'path/to/context' }}
      dockerfile: ${{ flow.workspace / 'path/to/context' }}
 ...  # Rest is same as above
```
It will automatically upload data to storage under some semi-random location (inside local action), and then run the remote build step and then return an image ref. It will be able to even utilize ``hash_files`` to avoid unnecessary reuploads and rebuilds.

**Comment by romasku** [comment]
After discussion in slack, we came up with the following solution:

We should add top-level property `images` to the batch flow and batch actions config file. This field will contain the following properties:
- `images.<image-name>.ref` -- image reference. Can use only locally accessible functions (such as `hash_files`). Its value will be calculated before the remote executor starts.
- `images.<image-name>.context` -- context for building an image. Can be either local path (e.g. `${{ flow.workspace / 'some/dir' }}`) or a remote volume spec (e.g. `storage:subdir/${{ inputs.path_on_storage }}`). If it is a local path, it cannot use anything that is not available at the beginning of a bake (for example, action inputs). If it is a remote path, usage of dynamic values is allowed. Local context should be automatically uploaded to storage during the "local actions" step of the bake.
- `images.<image-name>.dockerfile` -- path to a docker file. Work almost the same as `.context` - if it is a local path, the dynamic values are forbidden and it is automatically uploaded. If it a remote path, then dynamic values are allowed.
- `images.<image-name>.build_args`, `images.<image-name>.env`, `images.<image-name>.voluems` -- should work the same as in `live` mode. Should support dynamic (actions `inputs`) values.
- `images.<image-name>.force_rebuild` -- controls rebuilds between different bakes.

When a remote executor starts a new task, it should compare its `image` value with visible `image.<image-name>.ref` as strings. Visibility is defined in the following way: action sees images defined by itself, by parent action, by parent's parent action and so force up to the main config. For the main config, only top-level images are visible. If the task's `.image` value doesn't match any image, the executor tries to use it as is. If there was a match, then the following algorithm should be applied:
```
┌──────────────────────────┐
│ Is the corresponding ref │ Yes
│ available under neuro    ├─────────────┐
│      registry?           │             │
└──────────┬───────────────┘             │
           │                             │
           │No                 ┌─────────▼──────────┐
           │               Yes │    Is the flag     │
┌──────────▼───────────┐  ┌────┤ force_rebuild set? │
│  Is this image is    │  │    └─────────┬──────────┘
│ already being build  │  │              │
│  for some other task ◄──┘              │No
│    in this bake?     │                 │
└───┬─────────────┬────┘       ┌─────────▼──────────┐
    │             │            │ Use the image from │
    │No           │Yes         │    the registry    │
    │             │            └────────────────────┘
    │             │
    │             └─────────────────┐
    │                               │
┌───▼───────────────┐   ┌───────────▼──────────────┐
│ Start building of │   │ Wait until that image is │
│     the image     │   │ built and the reuse it   │
└───────────────────┘   └──────────────────────────┘
```
If task execution is blocked by image build, it should receive a "`Building`" state. If the build fails, the corresponding task should fail.
There should be a way to see the list of "builder" jobs associated with a given bake, so if some of them fail, the user will have a way to check the logs.

**Comment by YevheniiSemendiak** [comment]
@romasku, man, the solution looks awesome!
Thank you a lot for the summary and visualization (btw, which tool did you use?)

Some minor concerns are left unclear for me:
- `images.<image-name>.ref` still could read public images from other registries (DockerHub, GCR, etc.), right?
- `images.<image-name>.context`, `Local context should be automatically uploaded to storage during the "local actions" step of the bake.` - it will be uploaded and linked to the build job automatically, right? So when the local actions are finished, everything should be running remotely and you could detach from the executor.

**Comment by romasku** [comment]
> `images.<image-name>.ref` still could read public images from other registries (DockerHub, GCR, etc.), right?

Yes, sure. But in this case, all other attributes are forbidden. So if the ref doesn't start with `image:`, then it is an error to include any other attribute.
>  ... it will be uploaded and linked to the build job automatically, right? So when the local actions are finished, everything should be running remotely and you could detach from the executor.

Yes, exactly. Neuro-flow will upload all files under `images.<image-name>.context` and dockerfile from `images.<image-name>.dockerfile` to storage before starting remote executor. So that if the remote executor is started, it is safe to detach. When it will be time to build an image, the executor will run `neuro-extras image build` with context and dockerfile arguments pointing to files uploaded to storage.
> (btw, which tool did you use?)

https://asciiflow.com/#/

**Comment by mariyadavydova** [comment]
This is a great summarization of our discussion, thank you, @romasku!

**Comment by YevheniiSemendiak** [comment]
Awesome, thank you, @romasku!

**Comment by romasku** [comment]
Closed by https://github.com/neuro-inc/neuro-flow/pull/412
